pkg/pubsubsource/pubsub_test.go

```diff
@@ -154,68 +156,83 @@ func TestMain(m *testing.M) {
-	//doneCh := make(chan struct{})
-	sub := pubsubClient.Subscription(subscriptionID)
-	pubsubsource := NewPubSubSource(pubsubClient, sub)
-	ctx := context.Background()
+
+	firstReadCtx, cancelFirstRead := context.WithCancel(context.Background())
+
+	subscription := pubsubClient.Subscription(subscriptionID)
+	pubsubSource := NewPubSubSource(pubsubClient, subscription)
+
+	firstSendDoneCh := make(chan struct{})
 	go func() {
-		<-time.After(5 * time.Second)
-		sendMessages(context.Background(), pubsubClient, TopicID, 5)
+		defer close(firstSendDoneCh) // Ensure to close channel after sending
+		time.Sleep(5 * time.Second)
+		sendMessages(firstReadCtx, pubsubClient, TopicID, 5)
 	}()
 
-	pubsubsource.Read(ctx, mocks.ReadRequest{
+	pubsubSource.Read(firstReadCtx, mocks.ReadRequest{
 		CountValue: 5,
 		Timeout:    20 * time.Second,
 	}, messageCh)
 	assert.Equal(t, 5, len(messageCh))
+	<-firstSendDoneCh
+	cancelFirstRead()
+
+	secondReadCtx, cancelSecondRead := context.WithCancel(context.Background())
 
 	// Try reading 4 more messages
 	// Since the previous batch didn't get acked, the data source shouldn't allow us to read more messages
 	// We should get 0 messages, meaning the channel only holds the previous 5 messages
-	// Creating a new subscriber
-	pubsubsource.Read(ctx, mocks.ReadRequest{
+	pubsubSource.Read(secondReadCtx, mocks.ReadRequest{
 		CountValue: 4,
 		Timeout:    10 * time.Second,
 	}, messageCh)
-	assert.Equal(t, 5, len(messageCh))
-
-	// Ack the first batch
-	// Acknowledge messages.
-	for i := 0; i < 5; i++ {
-		msg := <-messageCh
-		pubsubsource.Ack(ctx, mocks.TestAckRequest{
-			OffsetsValue: []sourcer.Offset{msg.Offset()},
-		})
-	}
+	assert.Equal(t, 5, len(messageCh)) // Expecting no new messages
+
+	// acknowledging the first batch
+	msg1 := <-messageCh
+	msg2 := <-messageCh
+	msg3 := <-messageCh
+	msg4 := <-messageCh
+	msg5 := <-messageCh
+	pubsubSource.Ack(secondReadCtx, mocks.TestAckRequest{
+		OffsetsValue: []sourcer.Offset{msg1.Offset(), msg2.Offset(), msg3.Offset(), msg4.Offset(), msg5.Offset()},
+	})
+	cancelSecondRead()
+
+	thirdReadCtx, cancelThirdRead := context.WithCancel(context.Background())
+	thirdSendDoneCh := make(chan struct{})
 	go func() {
-		<-time.After(5 * time.Second)
-		sendMessages(context.Background(), pubsubClient, TopicID, 6)
-
+		defer close(thirdSendDoneCh) // Ensure to close channel after sending
+		time.Sleep(5 * time.Second)
+		sendMessages(thirdReadCtx, pubsubClient, TopicID, 6)
 	}()
 
-	// Try reading 6 more messages
-	// Since the previous batch got acked, the data source should allow us to read more messages
-	// We should get 6 messages
-	sub2 := pubsubClient.Subscription(subscriptionID)
-	pubsubsource2 := NewPubSubSource(pubsubClient, sub2)
-	pubsubsource2.Read(ctx, mocks.ReadRequest{
+	// Getting a new subscriber from the existing subscription
+	// As previous subscribers are closed
+	pubsubSource.SetSubscription(pubsubClient.Subscription(subscriptionID))
```

Comment:
Why do we need to close the previous subscriber? In an active source, what's the effect of resetting a subscriber?

Reply:
Because Read function has context Timeout and its automatically closing the resources